refactor(nuxt): enhance `useFetch` and `useLazyFetch` request type

didavid61202 opened the pull request:
## Updates
- awaiting fix in nitro for `NitroFetchRequest` type to support string type (unjs/nitro#225)
_update at 2022/5/11_

- updated this PR according to `unjs/nitro`'s new `NitroFetchRequest`  type (unjs/nitro#208)
_update at 2022/5/11_

- Implementing this enhancement in the `unjs/nitro` repo for `$fetch` first (unjs/nitro#208), and will update this PR accordingly.
_update at 2022/5/7_

---

Enhance the useFetch and useLazyFetch request's type by inferring nitropack's server route's string literal (InternalApi interface). supporting auto-complete and showing type hints.

### 🔗 Linked issue
Resolve #4823

### ❓ Type of change

<!-- What types of changes does your code introduce? Put an `x` in all the boxes that apply. -->

- [x] 📖 Documentation (updates to the documentation or readme)
- [ ] 🐞 Bug fix (a non-breaking change that fixes an issue)
- [x] 👌 Enhancement (improving an existing functionality like performance)
- [x] ✨ New feature (a non-breaking change that adds functionality)
- [ ] ⚠️ Breaking change (fix or feature that would cause existing functionality to change)

### 📚 Description

Enhance the `useFetch` and `useLazyFetch` request's type by inferring nitropack's server route's string literal (InternalApi interface).

This will provide better DX and prevent typos while using `useFetch` and `useLazyFetch` by supporting auto-complete and showing type hints when entering the request. And should still be able to pass other valid request URLs like `https://example.com/api` or Request object like `new Request('/api/test')`

###  Usage scenario
assume project contains `server/api/test.ts` 
```ts
import type { NitroFetchRequest } from 'nitropack'

const { data } = useFetch(/** IDE should show hint: '/api/test' */)

const request: NitroFetchRequest = /** IDE should show hint: '/api/test' */
```

```ts
useFetch('/api/test') // ok
useFetch('https://example.com/api') // ok
useFetch(new Request('api/test')) // ok

useFetch(/** will not show internal routes ex: '__nuxt_error' */)
```

### 📝 Checklist

<!-- Put an `x` in all the boxes that apply. -->
<!-- If your change requires a documentation PR, please link it appropriately -->
<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [x] I have linked an issue or discussion.
- [x] I have updated the documentation accordingly.

Rigo-m commented:
One thing that's missing is the ability to type POST request body, is there a way to implement it?

didavid61202 commented:
> One thing that's missing is the ability to type POST request body, is there a way to implement it?

@Rigo-m I'll look into this!

Rigo-m commented:
Sorry that I can't help, I'm still bad a TS somehow hahaha

didavid61202 commented:
> Sorry that I can't help, I'm still bad a TS somehow hahaha

Got a working code that infers the body type of a POST API, but still needs to manually export the type from the API file and change the auto-generated InternalApi interface, or have to make some change to nitropack (and maybe h3) to make the implementation cleaner.
I would have to discuss this with @danielroe or other core team members about this for code organization and naming consistency.

Rigo-m commented:
Maybe a clean solution could be manually typing body by extending nuxt schema like we do for manually tiping useRuntimeConfig?

pi0 commented:
I also love to try this idea! Indeed nitro seems better place to implement types for `$fetch`.

didavid61202 commented:
> I love this idea. I think rather than seeking to implement within `useFetch` within Nuxt we should do so within nitropack which is where the rest of the `$fetch` inference comes from: https://github.com/unjs/nitro/blob/4fbeef9bfd5c8e064817865a31d61927addc5e2a/src/types/fetch.ts
> 
> ... and then update this PR accordingly.
> 
> WDYT @pi0?

Sounds great! I implement it in Nuxt first as I wasn't sure if this should be implemented in nitropack. But after some digging and trying to implement type hint for POST body yesterday, I also found out that the codes to updates are mostly in nitropack.

mainly:
1. move this PR's new FetchRequestUrl type into nitropack
2. the structure of the `InternalApi` interface: except for the default return of server function, we can also expose the request's body type...etc as a Map type for future expansion? (as name export in server files or update `defineEventHandlers` signature)

I'll find some time to dig into it 👍

didavid61202 commented:
> > I love this idea. I think rather than seeking to implement within `useFetch` within Nuxt we should do so within nitropack which is where the rest of the `$fetch` inference comes from: https://github.com/unjs/nitro/blob/4fbeef9bfd5c8e064817865a31d61927addc5e2a/src/types/fetch.ts
> > ... and then update this PR accordingly.
> > WDYT @pi0?
> 
> Sounds great! I implement it in Nuxt first as I wasn't sure if this should be implemented in nitropack. But after some digging and trying to implement type hint for POST body yesterday, I also found out that the codes to updates are mostly in nitropack.
> 
> mainly:
> 
> 1. move this PR's new FetchRequestUrl type into nitropack
> 2. the structure of the `InternalApi` interface: except for the default return of server function, we can also expose the request's body type...etc as a Map type for future expansion? (as name export in server files or update `defineEventHandlers` signature)
> 
> I'll find some time to dig into it 👍

Hey, @danielroe  @pi0 would like to get some suggestions/notices on point 2 before I start, for changing the `InternalApi` interface and `defineEventHandlers` signature.
And if we want to auto infer Body type by using `useBody`, we might also have to update h3?
Thanks in advance ❤️

didavid61202 commented:
### Update 
Implementing this enhancement in `unjs/nitro` repo first (unjs/nitro#208), and will update this PR accordingly.

pi0 commented:
@didavid61202 Since https://github.com/unjs/nitro/issues/208 is landed, do you mind to rework this PR? Thanks!

didavid61202 commented:
I'm on it! 🚀

didavid61202 commented:
refactored and implemented with the new `NitroFetchRequest ` type from `nitro`, @danielroe @pi0  please review if you have time. 😉 

btw, I'm currently working on implementing type defining / inferencing for the event's method, body, and query (mainly with a new `definedContext` method). I will start at h3 > nitro > nuxt 3, wonder if this would interfere with any of the original road maps? Thanks!
  
### Update
update PR  #4825 according to `unjs/nitro`'s new `NitroFetchRequest ` type (unjs/nitro#208)
_update at 2022/5/11_

danielroe commented on docs/content/3.api/1.composables/use-fetch.md:
````diff
@@ -1,12 +1,12 @@
 ```ts [Signature]
 function useFetch(
-  url: string | Request,
+  url: NitroFetchRequest,
````
I think it's more useful for users to keep it as is:
```suggestion
  url: string | Request,
```

danielroe commented on docs/content/2.guide/2.features/5.data-fetching.md:
````diff
@@ -73,6 +73,7 @@ This composable provides a convenient wrapper around `useAsyncData` and `$fetch`
 
 ```vue [app.vue]
 <script setup>
+// will show '/api/count' as type hint, assume project contains server/api/count.ts
````
```suggestion
```

danielroe commented on docs/content/2.guide/2.features/5.data-fetching.md:
```diff
@@ -106,6 +107,7 @@ This composable behaves identically to `useFetch` with the `lazy: true` option s
 </template>
 
 <script setup>
+// will show '/api/posts' as type hint, assume project contains server/api/posts.ts
```
```suggestion
```

danielroe commented on docs/content/3.api/1.composables/use-lazy-fetch.md:
```diff
@@ -28,6 +28,7 @@ By default, [useFetch](/api/composables/use-fetch) blocks navigation until its a
 /* Navigation will occur before fetching is complete.
   Handle pending and error states directly within your component's template
 */
+// will show '/api/posts' as type hint, assume project contains server/api/posts.ts
```
```suggestion
```

danielroe reviewed (commented):
Some small comments.

Right now the big issue with this is that users can't do:

```js
const someDynamicUrl: string = 'https://test.com'
$fetch(someDynamicUrl)
```

We either need to document that fully or provide a way that this can work. Any ideas?

didavid61202 commented:
> Some small comments.
> 
> Right now the big issue with this is that users can't do:
> 
> ```js
> const someDynamicUrl: string = 'https://test.com'
> $fetch(someDynamicUrl)
> ```
> 
> We either need to document that fully or provide a way that this can work. Any ideas?

Thanks for pointing that out. That's indeed a big issue!

This should be fix by adding ‵string & {} to `NitroFetchRequest ` in nitro as:
`type NitroFetchRequest = keyof InternalApi | (`${string}${'/'}${string}` & {}) | Exclude<FetchRequest, string> | string & {}`

the only small downside is losing type hint when quotes are already typed,
$fetch( /** cursor here */ ) // shows routes hints
$fetch('/** cursor here */') // won't show hints, might have to tap into typescript hints using extendtion to support this?

@danielroe  @pi0  I'll work on a fix PR in nitro repo first, and inform you in this PR.

didavid61202 commented:
@pi0 I saw that type `NitroFetchRequest` was imported and partially implemented in `fetch.ts` by 7e89fe8, I merged the latest main into this PR and will update the rest accordingly.

didavid61202 commented:
Hi @pi0 @danielroe, I've updated this PR with the latest fix in nitro `NitroFetchRequest` for the `string` type as input. Please review if you got time, thanks 💚 

and I was wondering if we should exclude the `/__nuxt_error` path from `NitroFetchRequest`, or still let user sees it?
maybe we could also rename and export the Request type by `FetchRequest` or `NuxtFetchRequest`, any suggestion? 😃 

```ts
export type FetchRequest = Exclude<NitroFetchRequest, '/__nuxt_error'>
// or
// export type NuxtFetchRequest = Exclude<NitroFetchRequest, '/__nuxt_error'>
```

pi0 commented:
> and I was wondering if we should exclude the /__nuxt_error path from NitroFetchRequest, or still let user sees it?
maybe we could also rename and export the Request type by FetchRequest or NuxtFetchRequest, any suggestion? 😃

I would hide all `/_*` and `/api/_*` routes from auto completion as they are internal.

didavid61202 commented:
> > and I was wondering if we should exclude the /__nuxt_error path from NitroFetchRequest, or still let user sees it?
> > maybe we could also rename and export the Request type by FetchRequest or NuxtFetchRequest, any suggestion? 😃
> 
> I would hide all `/_*` and `/api/_*` routes from auto completion as they are internal.

updated in [b49385c](https://github.com/nuxt/framework/pull/4825/commits/b49385cf9bd86748349dd6cfd252c21a4b6cbfad), excluding all  `/_*` and `/api/_*` routes.

pi0 commented:
@[didavid61202](https://github.com/didavid61202) Maybe we can move this fix directly to Nitro and avoid workaround? I can quickly merge it for nitro and this after that :)

didavid61202 commented:
> @[didavid61202](https://github.com/didavid61202) Maybe we can move this fix directly to Nitro and avoid workaround? I can quickly merge it for nitro and this after that :)

sure, on it!

danielroe commented:
Looking great! There were a couple of workarounds in https://github.com/nuxt/framework/commit/7e89fe8f6c9619b30064665e17e98cedaae18f04 that I think we can now remove as well, if you're up for it 😊

pi0 commented:
> @[didavid61202](https://github.com/didavid61202) Maybe we can move this fix directly to Nitro and avoid workaround? I can quickly merge it for nitro and this after that :)

sure, on it!
and how is the alias's nameing? adding Nuxt prefix? or keep it as FetchRequest?

Can we directly use `NitroFetchRequest` everywhere? It reduces chances of inconsistent typings.

didavid61202 commented:
> 

@danielroe hi, you mean those at `fetch.ts` line 33,34 and 38 
`let r = request as Ref<FetchRequest> | FetchRequest | (() => FetchRequest)`
`return (isRef(r) ? r.value : r) as NitroFetchRequest`
and also in `error.ts` ?

didavid61202 commented:
> > and how is the alias's nameing? adding Nuxt prefix? or keep it as FetchRequest?
> 
> Can we directly use `NitroFetchRequest` everywhere? It reduces chances of inconsistent typings.

got it!

danielroe commented on docs/content/2.guide/2.features/5.data-fetching.md:
```diff
@@ -64,7 +64,7 @@ watch(count, (newCount) => {
 Within your pages, components and plugins you can use `useFetch` to universally fetch from any URL.
 
-This composable provides a convenient wrapper around `useAsyncData` and `$fetch`. It automatically generates a key based on URL and fetch options, as well as infers API response type.
+This composable provides a convenient wrapper around `useAsyncData` and `$fetch`. It automatically generates a key based on URL and fetch options, provide type hints for request url base on server routes, as well as infers API response type.
```
```suggestion
This composable provides a convenient wrapper around `useAsyncData` and `$fetch`. It automatically generates a key based on URL and fetch options, provides type hints for request url based on server routes, and infers API response type.
```

pi0 commented:
Thanks for working on this @didavid61202 @danielroe <3
